### PR TITLE
[ABO-237] Affiliate Pallet

### PR DESCRIPTION
**Pull request by DidacSF**
## Description

See [here](https://ajunanetwork.atlassian.net/wiki/spaces/PLATFORM/pages/206471169/Affiliate+System) for details.

## Type of changes

- [ ] `build`: Changes that affect the build system or external dependencies (eg, Cargo, Docker)
- [ ] `ci`: Changes to CI configuration
- [ ] `docs`: Changes to documentation only
- [x] `feat`: Changes to add a new feature
- [ ] `fix`: Changes to fix a bug
- [ ] `refactor`: Changes that do not alter functionality
- [ ] `style`: Changes to format the code
- [ ] `test`: Changes to add missing tests or correct existing tests

## Checklist

- [ ] Tests for the changes have been added
- [ ] Necessary documentation is added (if appropriate)
- [ ] Formatted with `cargo fmt --all`
- [ ] Linted with `cargo clippy --all-features --all-targets`
- [ ] Tested with `cargo test --workspace --all-features --all-targets`

### CODE REVIEW

File: pallets/ajuna-affiliate/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+		type AffiliateMaxLevel: Get<u32>;
+	}
+
+	#[pallet::storage]
+	pub type Organizer<T: Config> = StorageValue<_, T::AccountId, OptionQuery>;
+
+	/// Stores the affiliated accounts from the perspectives of the affiliatee
+	#[pallet::storage]
+	#[pallet::getter(fn affiliatees)]
+	pub type Affiliatees<T: Config> =
```
Comment (darkfriend77) [inline pallets/ajuna-affiliate/src/lib.rs]: I think we should use more common names.

**LinkedAccounts**

File: pallets/ajuna-affiliate/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		StorageMap<_, Identity, T::AccountId, AffiliatedAccountsOf<T>, OptionQuery>;
+
+	/// Stores the affiliated accounts from the perspective of the affiliator
+	#[pallet::storage]
+	#[pallet::getter(fn affiliators)]
+	pub type Affiliators<T: Config> = StorageMap<_, Identity, T::AccountId, bool, ValueQuery>;
```
Comment (darkfriend77) [inline pallets/ajuna-affiliate/src/lib.rs]: The comment seems wrong here.
Also, the naming should be more meaningful, **LinkableAccounts**

File: pallets/ajuna-affiliate/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+	/// Stores the affiliated accounts from the perspective of the affiliator
+	#[pallet::storage]
+	pub type Affiliatables<T: Config> = StorageMap<_, Identity, T::AccountId, (), OptionQuery>;
```
Comment (darkfriend77) [inline pallets/ajuna-affiliate/src/lib.rs]: This is unused I suppose, right?

File: pallets/ajuna-affiliate/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+		StorageMap<_, Blake2_128Concat, ExtrinsicId, u8, OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An organizer has been set.
+		OrganizerSet {
+			organizer: T::AccountId,
+		},
+		AccountAffiliated {
```
Comment (darkfriend77) [inline pallets/ajuna-affiliate/src/lib.rs:92]: AccountLinked

File: pallets/ajuna-affiliate/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+			beneficiary: T::AccountId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// There is no account set as the organizer
+		OrganizerNotSet,
+		/// An account cannot affiliate itself
+		CannotAffiliateSelf,
```
Comment (darkfriend77) [inline pallets/ajuna-affiliate/src/lib.rs:109]: Good catch, would replace in general Affiliate with Link

File: pallets/ajuna-affiliate/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+			AffiliateRules::<T>::remove(extrinsic_id);
+			Self::deposit_event(Event::RuleCleared { extrinsic_id });
+
+			Ok(())
+		}
+
+		#[pallet::call_index(3)]
+		#[pallet::weight({10_000})]
+		pub fn affiliate_to(origin: OriginFor<T>, affiliator: T::AccountId) -> DispatchResult {
+			let affiliatee = ensure_signed(origin)?;
```
Comment (darkfriend77) [inline pallets/ajuna-affiliate/src/lib.rs]: using the affiliatee name is inconsistent with  Affiliatees storage, we should make sure that the naming creates a certain consistency,

using child or parent or ... even more specific names.

just account for affiliatee.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+			Self::deposit_event(Event::AccountAffiliated { account: affiliatee, to: affiliator });
+
+			Ok(())
+		}
+
+		#[pallet::call_index(4)]
+		#[pallet::weight({10_000})]
+		pub fn clear_affiliation(origin: OriginFor<T>) -> DispatchResult {
+			let affiliatee = ensure_signed(origin)?;
+
```
Comment (darkfriend77) [inline pallets/ajuna-affiliate/src/lib.rs]: We should add comment, why we don't change anything else then removing the linked accounts.

**Review by darkfriend77 — CHANGES_REQUESTED** [review]
besides the naming, which I think needs to be more precise, it's good.

File: pallets/ajuna-affiliate/src/lib.rs (hunk 3 above)
Comment (darkfriend77) [reply pallets/ajuna-affiliate/src/lib.rs]: We can use the  Affiliators for this purpose, an Account that get's added will be added with false, and as soon as he gets an account linked to it turns to true.

File: Cargo.toml [hunk 8]
```diff
@@ -15,6 +15,7 @@ members = [
     "pallets/ajuna-nft-staking/benchmarking",
     "primitives",
 ]
+resolver = "2"
```
Comment (clangenb) [inline Cargo.toml]: Very nice!

I am personally very picky, though, when it comes to structuring things. I prefer if these one-liner configs are on the top of the toml, so that they are more visible. Does this make sense to you?

File: pallets/ajuna-affiliate/README.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+# Ajuna Affiliate Pallet
+
+This pallet is used to link specific calls from other pallets to an affiliate
+program in which one account can refer others. This referral then helps the referrer by
+receiving a part of the fees that would go to the treasury, thereby
+scientizing people to invite other people.
```
Comment (clangenb) [inline pallets/ajuna-affiliate/README.md]: I think there is an autocorrect error. Did you want to say 'incentivizing'?

File: pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/mod.rs [hunk 10]
```diff
@@ -1,3 +1,5 @@
+#![allow(unused_imports)]
```
Comment (clangenb) [inline pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/mod.rs:1]: Isn't it better to fix this?

File: pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/combinator/mate.rs [hunk 11]
```diff
@@ -118,7 +118,7 @@ impl<T: Config> AvatarCombinator<T> {
 			})
 			.transpose()?;
 
-		Ok((leader_output, other_output.into_iter().chain(additional_output.into_iter()).collect()))
+		Ok((leader_output, other_output.into_iter().chain(additional_output).collect()))
```
Comment (clangenb) [inline pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/combinator/mate.rs:121]: Another thing, I am quite picky with: feature PRs should not implement other unrelated fixes. Especially for me who is new to the code base, this makes my life easier. :)

File: pallets/ajuna-affiliate/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The maximum depth of the affiliate relation chain,
+		#[pallet::constant]
+		type AffiliateMaxLevel: Get<u32>;
+	}
+
+	#[pallet::storage]
+	pub type Organizer<T: Config> = StorageValue<_, T::AccountId, OptionQuery>;
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: We might want to have a genesis config value for this, to simplify setting up test networks. What do you think?

File: pallets/ajuna-affiliate/src/lib.rs (hunk 1 above)
Comment (clangenb) [reply pallets/ajuna-affiliate/src/lib.rs]: I like that `Affiliatees` implies a direction, while `LinkedAccounts` does not. However, I do agree that the names are quite confusing, and even if `Affiliatee` has a notion of direction, it is undefined by the terminology who earns the rewards. If possible, we should find a term that defines both or add this extra part in the documentation.

Haha, this reminds me again that there are only two hard software problems: 1. cache invalidation, 2. variable naming.

File: pallets/ajuna-affiliate/src/lib.rs (hunk 12 above)
Comment (clangenb) [reply pallets/ajuna-affiliate/src/lib.rs]: I also think that `manager` or `admin` is a better fit for the name.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+			AffiliateRules::<T>::contains_key(extrinsic_id)
+		}
+
+		fn execute_rule_for(
+			extrinsic_id: &ExtrinsicId,
+			_account: &T::AccountId,
+		) -> Result<(), DispatchError> {
+			let _rule = {
+				let maybe_rule = AffiliateRules::<T>::get(extrinsic_id);
+				ensure!(maybe_rule.is_some(), Error::<T>::ExtrinsicAlreadyHasRule);
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: I think this error is wrong. It should be `MissingRuleForExtrinsic`.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// There is no account set as the organizer
+		OrganizerNotSet,
+		/// An account cannot affiliate itself
+		CannotAffiliateSelf,
+		/// The account is not allowed to receive affiliates
+		CannotAffiliateToAccount,
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: I have been surprised by the `to` preposition here, I was sure that it isn't correct, while the internet tells me it exists, but it is very uncommon. I suggest using `with` instead of `to`, which is much more common.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+			extrinsic_id: ExtrinsicId,
+		},
+		RuleCleared {
+			extrinsic_id: ExtrinsicId,
+		},
+		RuleExecuted {
+			extrinsic_id: ExtrinsicId,
+			account: T::AccountId,
+			beneficiary: T::AccountId,
+		},
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: So, this event will be thrown whenever transaction fees are paid during an extrinsic with said id?

File: pallets/ajuna-affiliate/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+	#[pallet::storage]
+	pub type Organizer<T: Config> = StorageValue<_, T::AccountId, OptionQuery>;
+
+	/// Stores the affiliated accounts from the perspectives of the affiliatee
+	#[pallet::storage]
+	#[pallet::getter(fn affiliatees)]
+	pub type Affiliatees<T: Config> =
+		StorageMap<_, Identity, T::AccountId, AffiliatedAccountsOf<T>, OptionQuery>;
+
+	/// Stores the affiliated accounts from the perspective of the affiliator
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: I would suggest that the documentation is more elaborate here:

```
Store affiliators aka accounts that have affilatees and earn rewards from them.

Such accounts can't be affiliatees anymore.
```

File: pallets/ajuna-affiliate/src/lib.rs (hunk 6 above)
Comment (clangenb) [reply pallets/ajuna-affiliate/src/lib.rs]: I think the naming is correct, but this is proof that we need better documentation and maybe more meaningful names XD

File: pallets/ajuna-affiliate/src/lib.rs (hunk 3 above)
Comment (clangenb) [reply pallets/ajuna-affiliate/src/lib.rs]: I like this idea, but would prefer an enum here.

```rust
pub enum AffiliateState {
    NoAffiliatees
    HasAffiliatees // maybe even with amount ??
}
```

File: pallets/ajuna-affiliate/src/lib.rs (hunk 1 above)
Comment (clangenb) [reply pallets/ajuna-affiliate/src/lib.rs]: How about we name them: `Downline` and `Upline` accounts? I think this solves both of our issues:

* It is directional
* It uses the common affiliate marketing terms

So:

* Affiliatees -> DownlineAccounts
* Affiliators: can stay the same; these are accounts, which have downline accounts
* Affiliatables: As @darkfriend77 suggests, I think it is unneded too.

Then do some renaming based on this suggestion.

File: pallets/ajuna-affiliate/src/tests.rs [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+fn test_extrinsic_id() {
+	use frame_support::traits::GetCallIndex;
+
+	ExtBuilder::default().build().execute_with(|| {
+		use frame_support::traits::PalletInfoAccess;
+		let balance_index = Balances::index();
+		println!("pallet index is {:?}", balance_index);
+		let call_index =
+			pallet::Call::<Test>::add_rule_to { extrinsic_id: (0, 0), rule: 0 }.get_call_index();
+		println!("call index is {:?}", call_index);
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/tests.rs]: Don't we want to assert the id here?

File: pallets/ajuna-affiliate/src/tests.rs [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+	});
+}
+
+mod organizer {
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/tests.rs]: Oh, I like the module structuring, I never thought of this!

File: pallets/ajuna-affiliate/src/tests.rs [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+
+			// Fourth step on the chain BOB <- ALICE <- CHARLIE <- DAVE <- EDWARD
+			assert_ok!(Affiliates::affiliate_to(RuntimeOrigin::signed(EDWARD), DAVE));
+
+			System::assert_last_event(mock::RuntimeEvent::Affiliates(
+				crate::Event::AccountAffiliated { account: EDWARD, to: DAVE },
+			));
+
+			assert!(Affiliators::<Test>::get(DAVE));
+			assert_eq!(Affiliatees::<Test>::get(EDWARD), Some(bounded_vec![DAVE, CHARLIE]));
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/tests.rs]: Nice, I like the documentation in this test!

File: pallets/ajuna-affiliate/src/traits.rs [hunk 20]
```diff
@@ -0,0 +1,18 @@
+use sp_runtime::DispatchError;
+
+pub type PalletIndex = u32;
+pub type CallIndex = u8;
+
+pub type ExtrinsicId = (PalletIndex, CallIndex);
+
+pub trait AffiliateHandler<AccountId> {
+	fn add_account_to_affiliatables(account: AccountId);
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/traits.rs]: Nice, I like this trait. I make the pallet easily reusable in the treasury handler. However, I don't understand why  `add_account_to_affiliatables` is in this trait, shouldn't the trait provide read access only?

**Review by clangenb — COMMENTED** [review]
Nice, good job keeping the code fairly simple! I really think we have mostly on issue to solve: naming. Even though the naming is correct, it was really hard for me to keep track of which `Aff*` does what. I think I have added a good terminology suggestion in one of the comments, which we can refine.

File: Cargo.toml (hunk 8 above)
Comment (DidacSF) [reply Cargo.toml]: It has to be placed here unfortunately, since it is part of the `workspace` section, see here -> https://doc.rust-lang.org/cargo/reference/resolver.html#resolver-versions

File: pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/mod.rs (hunk 10 above)
Comment (DidacSF) [reply pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/mod.rs:1]: Yeah, the thing is, if I remove it and the fix the issues I get compilation errors, so this error is a false flag right now I believe.

File: pallets/ajuna-affiliate/README.md (hunk 9 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/README.md]: You are right!

File: pallets/ajuna-affiliate/src/lib.rs (hunk 13 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/lib.rs]: You are right!

File: pallets/ajuna-affiliate/src/lib.rs (hunk 15 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/lib.rs]: Yeah, that's the basic idea

File: pallets/ajuna-affiliate/src/lib.rs (hunk 16 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/lib.rs]: :+1:

File: pallets/ajuna-affiliate/src/tests.rs (hunk 17 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/tests.rs]: This test was just for having an idea on how would a pallet call the trait methods. It will be removed later

File: pallets/ajuna-affiliate/src/traits.rs (hunk 20 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/traits.rs]: This is because we want other pallets (aka games) to have the final say in to which accounts should be able to use the affiliate system, this is done so that we can tie the system to some gameplay mechanic in those other pallets make use of the affiliate system.

File: Cargo.toml (hunk 8 above)
Comment (clangenb) [reply Cargo.toml]: I meant I prefer it to be on top of the long list of the `members` entry.

File: pallets/ajuna-affiliate/src/traits.rs (hunk 20 above)
Comment (clangenb) [reply pallets/ajuna-affiliate/src/traits.rs]: Ok, I see this makes sense. However, I would argue that we might want to separate this into two traits then because the rewards' distribution process, in our case tx fee handler, only needs read access.

File: pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/mod.rs (hunk 10 above)
Comment (clangenb) [reply pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/mod.rs:1]: Did you have to fix the issues because you have bumped the rust-tool chain? Usually, I prefer if a version bump is done in a separate PR to not introduce more unrelated changes in the PR.

Comment (DidacSF) [reply pallets/ajuna-awesome-avatars/src/types/avatar/versions/v2/mod.rs:1]: Yeah, I did it because of a cargo update I ran that made some dependencies require it

File: Cargo.toml (hunk 8 above)
Comment (DidacSF) [reply Cargo.toml]: Ah, true, I will move it then!

File: pallets/ajuna-affiliate/src/lib.rs (hunk 12 above)
Comment (clangenb) [reply pallets/ajuna-affiliate/src/lib.rs]: I have understood in today's meeting that you also call it organizer in other pallets. So we'd better stay consistent here.

File: pallets/ajuna-affiliate/src/traits.rs [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+	#[default]
+	NonAffiliatable,
+	Affiliatable,
+	Blocked,
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, Debug, Default, Copy, Clone, PartialEq)]
+pub struct AffiliatorState {
+	pub(crate) status: AffiliatableStatus,
+	pub(crate) affiliates: u32,
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/traits.rs]: At one point in the future, it will be an annoyance that this is not public. I suggest making it completely public.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+	}
+
+	/// Stores the affiliated accounts from the perspectives of the affiliatee
+	#[pallet::storage]
+	#[pallet::getter(fn affiliatees)]
+	pub type Affiliatees<T: Config> =
+		StorageMap<_, Identity, T::AccountId, AffiliatedAccountsOf<T>, OptionQuery>;
+
+	/// Store affiliators aka accounts that have affilatees and earn rewards from them.
+	/// Such accounts can't be affiliatees anymore.
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs:74]: Happy about the comment. :)

File: pallets/ajuna-affiliate/src/lib.rs [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+		}
+
+		fn clear_affiliation_for(account: &AccountIdOf<T>) {
+			if let Some(mut affiliate_chain) = Affiliatees::<T>::take(account) {
+				if let Some(affiliator) = affiliate_chain.pop() {
+					Affiliators::<T>::mutate(&affiliator, |state| {
+						state.affiliates = state.affiliates.saturating_sub(1);
+					});
+				}
+			}
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: I usually prefer to remove nesting in such cases, but I guess this is opinionated. :)

```rust
			Affiliatees::<T>::take(account)
			   .and_then(|affiliate_chain| affiliate_chain.pop())
			   .map(|affiliator| 
					Affiliators::<T>::mutate(&affiliator, |state| {
						state.affiliates = state.affiliates.saturating_sub(1);
					});
			);
```

not 100% sure if my approach works as is, but with minor changes it should.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+	impl<T: Config> Pallet<T> {
+		fn add_new_affiliate_to(
+			affiliator: T::AccountId,
+			affiliatee: T::AccountId,
+		) -> DispatchResult {
+			let accounts = {
+				let mut accounts_vec = match Affiliatees::<T>::take(&affiliator) {
+					Some(accounts) => accounts,
+					None => AffiliatedAccountsOf::<T>::new(),
+				};
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: How about an `unwrap_or_default()`?

File: pallets/ajuna-affiliate/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+
+				Self::try_add_account_to(&mut accounts_vec, affiliator.clone())?;
+
+				accounts_vec
+			};
+
+			Affiliatees::<T>::insert(affiliatee, accounts);
+			Affiliators::<T>::mutate(&affiliator, |state| {
+				state.affiliates = state.affiliates.saturating_add(1);
+			});
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: If we want to an error upon overflow as discussed, we can choose the following code:

```rust
			Affiliators::<T>::try_mutate(&affiliator, |state| {
				state.affiliates = state.affiliates.checked_add(1).map_err(//todo: define error)
			})?;
```

File: pallets/ajuna-affiliate/src/lib.rs (hunk 23 above)
Comment (clangenb) [reply pallets/ajuna-affiliate/src/lib.rs]: As always in runtime code, I would actually go for the defensive variant here, one must never assume correctness of other parts of the code:

```rust
                  Affiliators::<T>::try_mutate(&affiliator, |state| {
						state.affiliates = state.affiliates.checked_sub(1).map_err(|_| {
                            log::error!("state.affiliates run into an underflow, this is a fatal bug in the pallet"
                            return Err("Some error");
                        });
					})?;
```

I know it is tedious and verbose, but unless explicitly proven one or two lines above, I strongly favor being defensive here.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+			}
+			accounts
+				.try_insert(0, account)
+				.map_err(|_| Error::<T>::CannotAffiliateMoreAccounts.into())
+		}
+	}
+
+	impl<T: Config> AffiliateInspector<AccountIdOf<T>> for Pallet<T> {
+		fn get_affiliator_chain_for(account: &AccountIdOf<T>) -> Option<Vec<AccountIdOf<T>>> {
+			Affiliatees::<T>::get(account).map(|accounts| accounts.to_vec())
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: `StorageMap::get()` returns owned values, but `to_vec` performs unfortunately `clone()`, so we can save `O(n)` if we replace `accounts.to_vec()` with `accounts.into_inner()`.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+			accounts: &mut AffiliatedAccountsOf<T>,
+			account: T::AccountId,
+		) -> DispatchResult {
+			if accounts.len() == T::AffiliateMaxLevel::get() as usize {
+				accounts.pop();
+			}
+			accounts
+				.try_insert(0, account)
+				.map_err(|_| Error::<T>::CannotAffiliateMoreAccounts.into())
+		}
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: Nice solution 👍

File: pallets/ajuna-affiliate/src/tests.rs [hunk 28]
```diff
@@ -0,0 +1,251 @@
...
+					rule
+				),
+				Error::<Test>::ExtrinsicAlreadyHasRule
+			);
+		})
+	}
+}
+
+mod clear_rule {
+	use super::*;
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/tests.rs:52]: I think we miss the test of removing a nonexistent rule.

File: pallets/ajuna-affiliate/src/tests.rs [hunk 29]
```diff
@@ -0,0 +1,251 @@
...
+				<Affiliates as AffiliateMutator<AccountIdOf<Test>>>::try_add_affiliate_to(
+					&CHARLIE, &ALICE
+				),
+				Error::<Test>::CannotAffiliateAlreadyAffiliatedAccount
+			);
+		});
+	}
+
+	#[test]
+	fn affiliate_to_rejects_with_unaffiliable_account() {
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/tests.rs:256]: I guess we should also test that with a blocked account.

File: pallets/ajuna-affiliate/src/tests.rs [hunk 30]
```diff
@@ -0,0 +1,251 @@
...
+			);
+		});
+	}
+}
+
+mod clear_affiliation {
+	use super::*;
+
+	#[test]
+	fn clear_affiliation_should_work() {
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/tests.rs:288]: Maybe we should test `clear_affiliation_rejects_if_no_affiliation_exists`.

**Review by clangenb — COMMENTED** [review]
Very nice cleanup, thanks for removing the unrelated changes! I just noted a few details, but nothing major.

File: pallets/ajuna-affiliate/src/lib.rs (hunk 26 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/lib.rs]: Oh, didn't know that! I'll make the change!

File: pallets/ajuna-affiliate/src/tests.rs (hunk 28 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/tests.rs:49]: Hmm, well nothing happens if you do that in the current logic, so I didn't deem a test necessary

File: pallets/ajuna-affiliate/src/lib.rs (hunk 23 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/lib.rs]: These are very good suggestions! I'll get it done on the next commit!

File: pallets/ajuna-affiliate/src/tests.rs (hunk 30 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/tests.rs:288]: Hmm currently clearing your affiliation if you have none will not cause any error, but I can add a test with the current behaviour for now!

File: pallets/ajuna-affiliate/src/tests.rs [hunk 31]
```diff
@@ -0,0 +1,302 @@
...
+				)
+			);
+
+			assert_eq!(Affiliators::<Test>::get(BOB).affiliates, 0);
+			assert_eq!(Affiliatees::<Test>::get(ALICE), None);
+		});
+	}
+
+	#[test]
+	fn clear_affiliation_if_no_affiliation_exists() {
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/tests.rs]: Add the expected outcome to the name: `clear_affiliation_returns_ok_if_no_affiliation_exists` or maybe we can even use the frame supports `assert_noop!` macro?

**Review by clangenb — COMMENTED** [review]
Nice, I am happy that you liked my suggestions!

File: pallets/ajuna-affiliate/src/lib.rs [hunk 32]
```diff
@@ -0,0 +1,269 @@
...
+pub mod pallet {
+	use super::*;
+	use sp_runtime::ArithmeticError;
+
+	pub type AffiliatedAccountsOf<T, I> =
+		BoundedVec<<T as frame_system::Config>::AccountId, <T as Config<I>>::AffiliateMaxLevel>;
+
+	pub type PayoutRuleOf<T, I> = PayoutRule<<T as Config<I>>::AffiliateMaxLevel>;
+
+	pub type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: This is a nitpick, but substrate has introduced the convention of using `For` instead of `Of` in these type aliases. I usually prefer aligning those practices to help people who are used to substrate jargon.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 33]
```diff
@@ -0,0 +1,269 @@
...
+			affiliator: T::AccountId,
+			affiliatee: T::AccountId,
+		) -> DispatchResult {
+			let accounts = {
+				let mut accounts_vec = Affiliatees::<T, I>::take(&affiliator).unwrap_or_default();
+
+				Self::try_add_account_to(&mut accounts_vec, affiliator.clone())?;
+
+				accounts_vec
+			};
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: I think you can just remove the nesting here ;)

File: pallets/ajuna-affiliate/src/lib.rs [hunk 34]
```diff
@@ -0,0 +1,269 @@
...
+		fn try_add_affiliate_to(
+			account: &AccountIdOf<T>,
+			affiliate: &AccountIdOf<T>,
+		) -> DispatchResult {
+			ensure!(account != affiliate, Error::<T, I>::CannotAffiliateSelf);
+
+			let affiliate_state = Affiliators::<T, I>::get(affiliate);
+			ensure!(
+				affiliate_state.affiliates == 0,
+				Error::<T, I>::CannotAffiliateExistingAffiliator
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: This error is a bit too cryptic for me without preposition. So I would add either `AffiliateWith` or `AffiliateTo` existing affiliator.

File: pallets/ajuna-affiliate/src/lib.rs [hunk 35]
```diff
@@ -0,0 +1,269 @@
...
+
+			ensure!(
+				!Affiliatees::<T, I>::contains_key(affiliate),
+				Error::<T, I>::CannotAffiliateAlreadyAffiliatedAccount
+			);
+
+			let affiliator_state = Affiliators::<T, I>::get(account);
+			ensure!(
+				affiliator_state.status == AffiliatableStatus::Affiliatable,
+				Error::<T, I>::CannotAffiliateToAccount
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs]: Maybe it is better if we phrase it more coherent with the check: `Error::TargetAccountIsNotAffiliatable`

**Review by clangenb — COMMENTED** [review]
Nothing really to add, current state looks good. 👍

File: pallets/ajuna-awesome-avatars/benchmarking/src/lib.rs [hunk 36]
```diff
@@ -37,15 +37,15 @@ use sp_std::vec;
...
 type CurrencyOf<T> = <T as AvatarsConfig>::Currency;
 type KeyLimitOf<T> = <T as AvatarsConfig>::KeyLimit;
 type ValueLimitOf<T> = <T as AvatarsConfig>::ValueLimit;
 
 type CollectionIdOf<T> = <<T as AvatarsConfig>::NftHandler as NftHandler<
-	AccountIdOf<T>,
+	AccountIdFor<T>,
 	AvatarIdOf<T>,
 	KeyLimitOf<T>,
 	ValueLimitOf<T>,
```
Comment (clangenb) [inline pallets/ajuna-awesome-avatars/benchmarking/src/lib.rs:51]: Ooops, I see that this is now inconsistent with the rest of your repository. I suggest that we clean this up in a follow-up PR:

File: pallets/ajuna-affiliate/src/lib.rs [hunk 37]
```diff
@@ -0,0 +1,269 @@
...
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self, I>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The rule identifier type at runtime.
+		type RuleIdentifier: Parameter + MaxEncodedLen;
+
+		/// The rule type at runtime.
+		type RuntimeRule: Parameter + MaxEncodedLen;
```
Comment (clangenb) [inline pallets/ajuna-affiliate/src/lib.rs:60]: Yes, exactly, this is how I meant it. I have been thinking some more about a trait bound we could introduce here, like `ExecuteRule`. However, as this pallet has no predefined hooks, meaning the pallet that uses the affiliate pallet can decide when to use this it won't be needed.

However, then that pallet needs to invoke the execution of the rule itself, then the question is, why do we even store the rule here? It is not needed, the invoking pallet already knows, which rule it wants to execute. Then this pallet, would actually only need to handle affiliate relations and no rules at all, and the other pallet simply wants to check if an account is an affiliator or not.

Alternatively, if we really and this to be generic, then maybe the `RuleId` does not correspond to the extrinsic enum variant. This would allow some runtime to even have two pallets, one that wants to add affiliate relations, and another one that would want to execute the rule. In this case, we'd need to store the rule here, but we'd need to have a trait bound so that the other pallet could execute the rule, if it finds one.

This is probably one more fundamental thing we have to discuss in the brainstorming session tomorrow.

**Review by clangenb — COMMENTED** [review]
Nice, codewise nothing to add, I just have one fundamental question. :)

File: pallets/ajuna-affiliate/src/lib.rs (hunk 37 above)
Comment (DidacSF) [reply pallets/ajuna-affiliate/src/lib.rs:60]: Yeah, I get what you mean, it really is something that we should discuss in the meeting

**Review by darkfriend77 — APPROVED** [review]
